src/Data/Text/Encoding.hs

```diff
@@ -187,31 +233,274 @@ decodeLatin1 bs = withBS bs $ \fp len -> runST $ do
 foreign import ccall unsafe "_hs_text_is_ascii" c_is_ascii
     :: Ptr Word8 -> Ptr Word8 -> IO CSize
 
-isValidBS :: ByteString -> Bool
+-- | This data type represents the state of a 'ByteString' representing
+-- UTF-8-encoded text. It consists of a value representing whether or
+-- not the last byte is a complete code point, and on incompletion what
+-- the 1 to 3 end bytes are that make up the incomplete code point.
+data Utf8ValidState = Utf8ValidState
+  { -- | Get the incomplete UTF-8 code point of the 'ByteString's that
+    -- have been validated thus far.
+    partialUtf8CodePoint :: [ByteString]
```

Comment:
Good point.

Reply:
I started working on this, and I'm wondering if the juice is worth the squeeze.

In the current form, with the head bytestring a reference to it is already being maintained in the `TextDataStack` since `ByteString.drop` and `ByteString.take` simply move the needle without changing the underlying array. With any tail bytestrings, they'll only have a length of one or two words. (Admittedly, this all with the understanding that a bytestring is not just the underlying array, but also a pointer, an integer, and the structure binding the three together, and that each drop and take create more of these structures.)

Additionally, it adds an intermediary translation and the cost of going to and from that translation: `ByteString` -> `Word32` -> `Text`.

How much added value is this for at most three bytestrings whose total length is at most three words?